[LOOKML-4] LookML parser

julianhyde opened the pull request:
Fixes #4

julianhyde committed:
Add class PairList

PairList and related classes are copied from code that I
wrote in Calcite.

Add checker-qual and error_prone_annotations as dependencies.

Remove Lookml and LookmlTest classes. (They were
placeholders to allow the build to work in the initial
revision, and are no longer needed.)

julianhyde committed:
[LOOKML-2] Add interface ObjectHandler, an event model for well-formed LookML

Fixes lookml#2

julianhyde committed:
[LOOKML-4] LookML parser

Add a parser for well-formed LookML.

By "well-formed LookML" we mean syntactically valid LookML
strings. No schema is enforced; so, for example, the root
property does not need to be named "model", or "view"; it
can just as well be called "foo".

The parser is implemented in JavaCC (a Java-based parser
generator that generates top-down parsers). It is a push
parser (like SAX in the XML world) that does not build an
AST but just sends events to an ObjectHandler.

Fixes #4

julianhyde committed:
Documentation

julianhyde committed:
Documentation 2

julianhyde committed:
Move fields

amarkowitz commented on src/main/javacc/LookmlParser.jj:
```diff
@@ -0,0 +1,338 @@
+
+options {
+  STATIC = false;
+  IGNORE_CASE = true;
```
case ignored?

julianhyde replied:
Only relevant to keywords (e.g. `if`), and LookML has no keywords. Removed.

amarkowitz reviewed (approved):
Seems reasonable to me.  Biggest question I have is about using javacc vs antlr (same as a previous PR)

julianhyde commented on src/main/java/net/hydromatic/lookml/LaxParser.java:
```diff
@@ -0,0 +1,49 @@
+      Collection<String> codePropertyNames, String code) {
+    final Reader reader = new StringReader(code);
+    final LookmlParserImpl parser = new LookmlParserImpl(reader);
+    parser.setCodePropertyNames(ImmutableSortedSet.copyOf(codePropertyNames));
```
Unavoidable. `LookmlParserImpl` is generated by JavaCC, and it's not possible to add extra parameters to the constructors.

Best we can do is keep `LookmlParserImpl` private.

julianhyde replied:
I've made it package-private now.

julianhyde commented:
> Biggest question I have is about using javacc vs antlr (same as a previous PR)

As before, I think there are two questions:
1. Should the parser create an AST, or events?
2. JavaCC vs Antlr.

Question 2 is much less important than question 1, provided that the parser doesn't force us to create an AST. (Maybe one has better performance, or error messages, or a grammar that is easier to read, or a more convenient runtime library.) I used JavaCC just because I know it. I believe someone could - at any time we choose - swap it for Antlr with very little impact to the project.

julianhyde commented on README.md:
```diff
@@ -26,6 +26,176 @@ License.
+instance, as long as the names are unique.
+
+<!--
+To be decided (or documented):
```
I moved this whole comment to #18

julianhyde commented on README.md:
```diff
@@ -26,6 +26,176 @@ License.
+
+2. Must the root property be a named-object?
+
+3. 'List' or 'collection'?
```
agreed

julianhyde commented on README.md:
```diff
@@ -26,6 +26,176 @@ License.
+7. We said above that `founded` and `member` are property types.
+But in the API (e.g. `code(String propertyName, String value)` they are
+called property names. And in the schema, they are properties (short for
+property definitions). So should we call them property names?
```
formally there are 'property definition' and 'property instance', but informally we can use 'property' for either as long as it is clear. E.g. `instruments` has one definition (in the schema) and 4 instances (in this particular doc).

julianhyde commented on src/main/java/net/hydromatic/lookml/LaxParser.java:
```diff
@@ -0,0 +1,49 @@
+    } catch (ParseException e) {
+      throw new RuntimeException(e);
+    }
+    handler.close();
```
no, it's not that kind of `close`. it should only be called if parsing is successful. I moved the call to `close` inside the parser grammar.

julianhyde committed:
Changes based on Adam's review

Move LaxParser into the parser package (so that it can
access private methods of LookmlParserImpl) and rename
it to LookmlParsers.

Obsolete interface LookmlParser.

In README, clarify the difference between property
definition and property instance, and that both can be
referred to as property if unambiguous.

In LookmlParsers add interface Config, so we can add
parameters in future.

Remove parser IGNORE_CASE.

julianhyde commented on src/main/javacc/LookmlParser.jj:
```diff
@@ -0,0 +1,338 @@
+
+/* LITERALS */
+
+<DEFAULT> TOKEN :
```
Please log issues.

amarkowitz replied:
issue added : #25

julianhyde commented on src/main/java/net/hydromatic/lookml/parse/LookmlParser.java:
```diff
@@ -0,0 +1,29 @@
+/** Parser for LookML. */
+public interface LookmlParser {
+  /** Returns the position of the last token returned by the parser. */
+  Pos pos();
```
It was unnecessary. I removed it.

julianhyde commented on src/main/java/net/hydromatic/lookml/LaxParser.java:
```diff
@@ -0,0 +1,49 @@
+    try {
+      parser.document(handler);
+    } catch (ParseException e) {
+      throw new RuntimeException(e);
```
It's a checked exception, and IMHO checked exceptions add useless boilerplate to APIs.

We do need a better error handling strategy, but checked exceptions aren't how we get there.

julianhyde commented on src/main/java/net/hydromatic/lookml/Pos.java:
```diff
@@ -0,0 +1,243 @@
+  }
+
+  @Override public boolean equals(Object o) {
+    return o == this
```
fixed in the PR for #14

julianhyde commented on src/main/java/net/hydromatic/lookml/Pos.java:
```diff
@@ -0,0 +1,243 @@
+   * Combines an iterable of parser positions to create a position which spans
+   * from the beginning of the first to the end of the last.
+   */
+  public static Pos sum(Iterable<Pos> poses) {
```
Yeah, `sum` isn't great. I thought about `span` and `union`. But `sum` works well with `plus` used elsewhere, so I think I'll keep it.
